rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+support the latter case and verify client certificates at re-joining time.
+
+#### Alternative: Public Key Fingerprint
+
+One option is to modify the behaviour of `tbot` to persist and reuse the
+keypair across renewals. We could then use a fingerprint of the public key as a
+unique identifier of the Bot instance.
+
+This feels like a natural identifier. It avoids introducing a new attribute to
+certificates as the public key is already encoded within certificates. The
+nature of public-key cryptography also means that this provides the Bot instance
+a way to identify itself without needing an issued certificate.
+
+To mitigate the risk of pre-image attacks, SHA256 will be used to determine the
+fingerprint of the public key. In addition, the full public key should be
+recorded and verified against when authenticating a Bot instance action.
+
+It should be noted that with this technique, rotating the keypair of a `tbot`
+instance would reset the identity of that instance. Rotation of this keypair
+would be unusual and this side effect seems expected.
+
+This technique does have some downsides:
+
+- It makes it impossible to rotate a bot's private key. However, we do not
+  currently support this today, and purging a bot's data directory to do so
+  would simply result in a new `BotInstance`, which is likely an acceptable
+  workaround.
+- Our join process today is unable to cryptographically verify the public key
+  presented by a joining bot to ensure that particular keypair has been issued
+  an identity already. Clients can provide any public key they like, including
+  that of an existing bot.
+
+### BotInstance Resource
+
+The
+[Resource Guidelines RFD](https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md)
+will be followed.
+
+With a persistent identifier for a Bot instance established, we can now track
+information about a specific Bot instance server-side. In addition to providing
+a way to store a generation counter per instance, this yields other benefits:
+
+- Allows Bot instances to be viewed within the UI and CLI.
+- Allowing Bot instances to submit basic self-reported information about itself
+  and its host, e.g:
+  - `tbot` version
+  - Hostname, OS and OS version
+  - The configuration of `tbot`
+  - Health status
+- Record metadata from delegated joins to enrich information about the Bot.
+  E.g show the linked repository / CI run number
+- Billing based on Bot instances rather than Bots.
+
+Some of this information is known and verified by the server - for example, the
+certificate generation or the join metadata. Some of this information is
+self-reported and should not be trusted. The information from these two sources
+should be segregated to avoid confusion.
+
+BotInstance will be a new resource type introduced to track this information.
+
+```protobuf
+syntax = "proto3";
+
+import "teleport/header/v1/metadata.proto";
+
+// A BotInstance
+message BotInstance {
```

Comment:
Does this resource have any impact on our Terraform provider?

Reply:
I don't think so, bot instances will be created automatically when a bot joins. They're more for record keeping and visibility than something end users need to configure.

I could see exposing it in the future if we add some command and control functionality to `BotInstanceSpec`, though.